### PR TITLE
[shardformer]fix gpt2 double head

### PR DESCRIPTION
**Pull request by flybird11111**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

fix GPT2DoubleHeadsModel test failed.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: tests/test_shardformer/test_model/test_shard_gpt2.py [hunk 1]
```diff
@@ -136,14 +136,6 @@ def check_forward_backward(model_fn, data_gen_fn, output_transform_fn, loss_fn,
     'num_microbatches': 4,
     'enable_all_optimization': True,
     'use_lazy_init': True,
-    'enable_sequence_parallelism': True,
```
[the remark this answers is not in the document]
Comment (flybird11111) [reply tests/test_shardformer/test_model/test_shard_gpt2.py:139]: enable_all_optimization  will activate sequence parallelism.

File: colossalai/shardformer/modeling/gpt2.py [hunk 2]
```diff
@@ -89,13 +89,14 @@ def gpt2_model_forward(
...
-                token_type_ids = token_type_ids.view(-1, seq_length)
+                token_type_ids = token_type_ids.view(-1, input_shape[-1])
         else:
             if hidden_states is None:
                 raise ValueError("hidden_states shouldn't be None for stages other than the first stage.")
             input_shape = hidden_states.size()[:-1]
-            batch_size, seq_length = input_shape[0], input_shape[1]
+            batch_size = input_shape[0]
             device = hidden_states.device
+            hidden_states = hidden_states.view((-1,) + hidden_states.shape[-2:])
```
[the remark this answers is not in the document]
Comment (flybird11111) [reply colossalai/shardformer/modeling/gpt2.py:99]: If the hidden_states'dim equals 4, rearrange (b, c, q ,d) to (b*c, q, d)

**Review by Fridge003 — APPROVED** [review]

File: tests/test_booster/test_plugin/test_gemini_plugin.py [hunk 3]
```diff
@@ -86,7 +86,8 @@ def check_gemini_plugin(subset: str, init_method: str = 'none', early_stop: bool
                 'transformers_t5_encoder_model',    # does not support apex rmsnorm
                 'transformers_chatglm',
                 'transformers_sam',
-                'transformers_vit'
+                'transformers_vit',
+                'transformers_gpt_double_heads'
```
Comment (ver217) [inline tests/test_booster/test_plugin/test_gemini_plugin.py:90]: What's wrong with this?

Comment (flybird11111) [reply tests/test_booster/test_plugin/test_gemini_plugin.py:90]: The test failed to run this model with gemini and raise a error that "synchronization of gradients doesn't exit properly",

**Review by ver217 — APPROVED** [review]

**Pull request merged**